Closing the Farm Silver (#701)
cpinitiative/usaco-guide: solutions/USACO open16-closing-the-farm.mdx

---
id: usaco-644
source: USACO Silver 2016 US Open
title: Closing the Farm
author: Jesse Choe
---

[Official Analysis](http://www.usaco.org/current/data/sol_closing_silver_open16.html)

<LanguageSection>
<CPPSection>

> In this problem, we are asked to determine if all remaining barns are connected as the barns close one at a time. To check this, we can run a depth-first search (DFS) starting from the barn that will close last. To simulate the closure each barn, we can store a boolean array and set each barn in the iteration to be "closed". Then, we can run the DFS and check how many nodes have been visited. If all the unclosed nodes have been visited, then we print "YES"; otherwise, we print "NO".

Reviewer: ```suggestion
In this problem, we are asked to determine if all remaining barns are connected as the barns close one at a time. To check this, we can run a depth-first search (DFS) starting from the barn that will close last. To simulate the closure of each barn, we can store a boolean array and set each barn in the iteration to be "closed". Then, we can run the DFS and check how many nodes have been visited. If all the unclosed nodes have been visited, then we print "YES"; otherwise, we print "NO".
```

Author: Yeah, I fixed it. Can you review my new PR now? @frodakcin

Reviewer: You should re-request review. Also, do you prefer to stick with long longs?

Author: How do I re-request? Since I changed to int's, I don't really care tbh.

Reviewer: Top right. Click the button that looks like a cyclic arrow loop thing

---
id: usaco-644
source: USACO Silver 2016 US Open
title: Closing the Farm
author: Jesse Choe
---

[Official Analysis](http://www.usaco.org/current/data/sol_closing_silver_open16.html)

<LanguageSection>
<CPPSection>

> In this problem, we are asked to determine if each barn is connected after a certain barn is closed. To check this, we can run a depth-first search (DFS) algorithm

Reviewer: ```suggestion
In this problem, we are asked to determine if all remaining barns are connected as the barns close one at a time. To check this, we can run a depth-first search (DFS)
```

> on the last node to be closed, since each line asks whether the closure of the previous barn will allow connection between each of the unclosed nodes, making the last

Reviewer: ```suggestion
starting from the barn that will close last, since each line asks whether the closure of the previous barn will allow connection between each of the unclosed nodes, making the last
```

> closed barn irrelevant to the output. To close each barn, we can store a boolean array and set each closed barn to be true for each iteration. Then, we can run the

Reviewer: Why is the last closed barn "irrelevant to the output?" Also, I think it would be preferable to start a new sentence at the word "since." (Possibly rewrite this sentence)

> DFS and check how many nodes have been visited; if all the unclosed nodes have been visited, then we print "YES"; otherwise, we print "NO".

Reviewer: If it's one paragraph, put everything on the same line.

If it's multiple paragraphs, put an empty line separating each paragraph. 

In this case, I think you should put lines 13-16 on the same line.

#define ub(a, x) upper_bound(all(a), x) - begin(a)

void setIO(string fileName = ""){
	FastIO;
	if(len(fileName)){
		freopen((fileName+".in").c_str(), "r", stdin);
		freopen((fileName+".out").c_str(), "w", stdout);
	}
}

vl adj[3001], order(3001); ll nodes;
vb visited(3001), closed(3001);

// Store the global variables, including the number of nodes visited and the boolean closed array.

void DFS(ll node){
    if(visited[node] || closed[node]) return;
    // Increase the number of nodes if and only if they haven't been closed.
    nodes++;
    visited[node] = true;
    EACH(u, adj[node]){
        if(!visited[u]){
            DFS(u);
        }
    }
}

int main(){
    setIO("closing");
    ll n, m; cin >> n >> m;
    FOR(i, 0, m){
        ll a, b; cin >> a >> b;
        adj[a].pb(b);
        adj[b].pb(a);
    }
    FOR(i, 0, n){
        cin >> order[i];
    }
    // Check the first line which indicates whether the initial farm is fully connected.
    DFS(1);
    if(nodes == n) cout << "YES" << endl;
    else cout << "NO" << endl;
    FOR(i, 0, n-1){
        fill(all(visited), false);
        nodes = 0;
        closed[order[i]] = true;

>         DFS(order[n-1]);

Reviewer: ```suggestion
        DFS(order[n-1]); // DFS from the barn that will close last
```

	// Have all the unclosed barns been visited?
        (nodes == n-i-1) ? cout << "YES" << endl: cout << "NO" << endl;
    }
}

> ```

Reviewer: Can we make all variables integers? No reason to use `long long`'s if we don't have to.

title: Closing the Farm
author: Jesse Choe, Maggie Liu
---

[Official Analysis](http://www.usaco.org/current/data/sol_closing_silver_open16.html)

In this problem, we are asked to determine if all remaining barns are connected as the barns close one at a time. To check this, we can run a depth-first search (DFS) starting from the barn that will close last. To simulate the closure of each barn, we can store a boolean array and set each barn in the iteration to be "closed". Then, we can run the DFS and check how many nodes have been visited. If all the unclosed nodes have been visited, then we print "YES"; otherwise, we print "NO".

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;
using vb = vector<bool>;
using pl = pair<ll, ll>;
using vpl = vector<pl>;
using vc = vector<char>;
using vs = vector<string>;
using str = string;

#define FastIO ios_base::sync_with_stdio(0); cin.tie(0)
#define pb push_back
#define mp make_pair
#define all(x) begin(x), end(x)
#define sor(x) sort(all(x))
#define ins insert
#define f first
#define s second
#define sz(x) (ll)(x).size()
#define len(x) (ll)(x).length()

> #define FOR(i, a, b) for(ll i=a; i<b; ++i)

Reviewer 2: Why do you use 'long long' instead of 'int'?

Reviewer: Should we ask benq?

Author: I transitioned from int to long long all together cuz I hate when u get like WA for forgetting long longs. I don't wanna go throught the hustle of changing, so I'm using ll instead of int @tachithanhdanh

Reviewer 2: > I transitioned from int to long long all together cuz I hate when u get like WA for forgetting long longs. I don't wanna go throught the hustle of changing, so I'm using ll instead of int @tachithanhdanh

But is it necessary to convert the variables in the loop to `long long`?
Also, why do you have to use `long long` for `sz(x)` and `len(x)`, both of which will never exceed the size of `int`?

Reviewer: The problem is that ints do have a significant advantage over lls, usually a noticeable amount of memory & runtime. Being able to automatically recognize when it's necessary to ll's over ints is a good skill too (and also lets you recognize when int128/bigint is necessary). So asking you to fix this would potentially fall under "bad practice," but we should probably ask @bqi343 if it's reasonable enough.

Reviewer 2: and one thing to note is that `long long` variables tend to run much slower than that of `int` because they use twice as much memory as the `int` do.

Author: Wait, so how significant is the slow runtime for using `long long`?

Reviewer: By almost a factor of 2. (Possibly more, given that the skipped soln only ran for pretests)

https://codeforces.com/contest/1348/submission/78738103
https://codeforces.com/contest/1348/submission/78739532
